Introduce expression enumerators

qinheping opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

This PR include expression enumerator interface that allow us to specify regular tree grammars over `exprt` and enumerate expressions from the grammar. It will be used in the enumerative loop-invariant synthesizer.

We use a placeholder class to represent non-terminal symbols, enumerator classes to represent productions, and a factory class to represent grammars. A factory contain a map from placeholders to enumerators (non-terminal symbols to their productions).  To enumerate expressions in the grammar with a given size, we repeatably expand placeholders with their productions until the expected size is reached. And then all productions without placeholders are instantiated to `exprt` objects.

* `enumerator_baset`: base class of enumerators.
* `leaf_enumeratort`: enumerators that enumerate leaf expressions---complete `exprt` without placeholder.
* `non_leaf_enumeratort`: enumerators that enumerate tree expressions of form `op(a_1, ...,  a_n)` where `op` is its operator symbol, `n` is the arity of `op`, and `a_i` are expressions enumerated by the `i`-th sub-enumerator.
* `binary_functional_enumeratort :  non_leaf_enumeratort`: enumerators that enumerate trees with two children. It is optimized for boolean operators, comparison operators, and the operator `+`.
* `alternatives_enumeratort`: enumerators that enumerate expressions from the union of expressions enumerated by their sub-enumerators. For example, an enumerator for productions `-> S + 1 | 1` in the grammar `S -> S + 1 | 1` should enumerate the union of `-> S + 1` and `-> 1`.
* `recursive_enumerator_placeholdert`: placeholders that represent non-terminals in the grammar.
* `enumerator_factory_baset`: factories that maintain enumerators, placeholders, and the map from placeholders to  enumerators.
<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

jimgrundy commented:
@SaswatPadhi, can you provide a first review of this so that we are making good use of the time of the DiffBlue reviewers.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+
+#include "expr_enumerator.h"
+
+std::list<exprst> non_leaf_enumeratort::cartesian_product_of_enumerators(
```
Just checking: is a list-of-lists really the most suitable type here? Should it perhaps be a set/an unordered set (i.e., a flat, de-duplicated collection of expressions)?

qinheping replied:
Yes, it should be a set of ```exprst```.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+
+std::list<exprst> non_leaf_enumeratort::cartesian_product_of_enumerators(
+  const enumeratorst &enumerators,
+  unsigned int pos)
```
I'm highly suspicious of any use of `unsigned int`. Should this perhaps be `std::size_t`? There's hardly any reason to ever use `unsigned int` (its width is architecture-dependent and yet not wide enough to hold size-of-a-container values, and it's also not the smallest unsigned type). Applies across this commit.

qinheping replied:
Thank you! I replace ```unsinged int``` with ```std::size_``` across this commit.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+  unsigned int pos)
+{
+  std::list<exprst> result = std::list<exprst>();
+  unsigned int tuple_size = enumerators.size();
```
You definitively want `std::size_t` here.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+  const enumeratorst &enumerators,
+  unsigned int pos)
+{
+  std::list<exprst> result = std::list<exprst>();
```
No need for that assignment/right-hand side here.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+  std::list<exprst> result = std::list<exprst>();
+  unsigned int tuple_size = enumerators.size();
+
+  if(tuple_size == pos + 1)
```
Do you perhaps want `pos` to be an iterator instead of an index?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+  if(tuple_size == pos + 1)
+  {
+    /// Current position points to the last enumerator.
+    /// Let the last enumerator enumerates and return the result.
```
I'm struggling to parse this sentence.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+    {
+      exprst new_tuple = exprst();
+      new_tuple.push_back(e);
+      result.push_back(new_tuple);
```
Could be written as `result.emplace_back(e);`

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+      new_tuple.push_back(e);
+      result.push_back(new_tuple);
+    }
+    return result;
```
All branches do this, so perhaps just leave it until the very end of this function?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+  }
+  else
+  {
+    /// First compute the cartesian product of enumerators after `pos`.
```
I believe it's "Cartesian," i.e., with an uppercase "C." But also: what does it mean to compute the Cartesian product of a single list? Given a list (a, b, c), is it (a, b, c) x (a, b, c) == (a*a, b*a, c*a, a*b, b*b*, c*b, a*c, b*c, c*c)?

qinheping replied:
Yes, (a, b, c) x (a, b, c) == (aa, ba, ca, ab, bb, cb, ac, bc, cc) is correct. But (a, b, c) x (a, b, c) is the Cartesian product of two sets that are the same (they are both (a, b, c)), not the Cartesian product of a single set.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+      for(const auto &elem : (*front)->enumerate())
+      {
+        exprst new_tuple = exprst(sub_tuple);
+        new_tuple.push_front(elem);
```
Does the order matter?

qinheping replied:
Yes. 
An an example, if we want to enumerate expressions for a production rule ```E_1 - E_2``` where ```E_1``` and ```E_2``` are two enumerators. We need to enumerate ```e_1 - e_2``` for all pair ```(e_1, e_2)``` where ```e_1``` is an expression enumerated by ```E_1```, and ```e_2``` is an expression enumerated by ```E_2```. 
```(e_1, e_2)``` is an ordered set.
Note that the set of all such pairs ```(e_1, e_2)``` is the Cartesian set of enumerated expressions of ```E_1``` and ```E_2```.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+std::list<partitiont>
+non_leaf_enumeratort::get_partitions(unsigned int n, unsigned int k)
+{
+  INVARIANT(n >= k, "too less elements to be partitioned");
```
s/less/few/

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+
+  std::list<partitiont> result = std::list<partitiont>();
+
+  unsigned int *cuts = new unsigned int[k + 1];
```
This is never freed. (And I'm not sure it shouldn't just be a `std::vector`.)

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+      for(unsigned int i = rightmost - 1; i >= 1; i++)
+      {
+        cuts[i] = cuts[rightmost] - accum;
+        accum++;
```
This code could do with some comments.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,112 @@
+  else
+  {
+    leaf_enumeratort none_enumerator = leaf_enumeratort(exprst(), true);
+    actual_enumerator = &none_enumerator;
```
Storing the address of an object that will go out of scope will result in invalid memory accesses.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+    {
+      result.push_back(expr);
+    }
+    return result;
```
Why not just `return leaf_expr;`?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+  /// Enumerate expressions in the given list.
+  exprst enumerate() override
+  {
+    exprst result = exprst();
```
Can be dropped.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+
+    // Do nothing if the size of epxected enumerated expression is not 1.
+    if(expected_size != 1)
+      return result;
```
Just `return {};`

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+  {
+    exprst result = exprst();
+
+    // Do nothing if the size of epxected enumerated expression is not 1.
+    if(expected_size != 1)
```
Is this a temporary measure, or is this meant to stay?

qinheping replied:
This is meant to stay. It enumerate nothing when the assigned size is _bad_. For leaf enumerators, the only good size is one.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+  {
+  }
+
+  bool is_none_enumerator;
```
Should this be `const`?

qinheping replied:
Yes.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+
+#include "synthesizer_utils.h"
+
+#include <iostream>
```
Is this really needed/used?

qinheping replied:
No, it was for debugging only.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+class enumerator_baset
+{
+public:
+  explicit enumerator_baset(bool is_none_enumerator = false)
+    : is_none_enumerator(is_none_enumerator)
+  {
+  }
```
This needs a comment: when should the parameter be set to true/false?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+  virtual exprst enumerate() = 0;
+  virtual enumerator_baset *clone() const = 0;
+
+  virtual ~enumerator_baset()
+  {
+  }
```
You might use `= default;`

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+protected:
+};
+
+typedef std::list<enumerator_baset *> enumeratorst;
```
At this point I'm really asking myself: what is the overall architecture? I'm afraid I'll need some file-level comment to explain 1) how to use this and 2) what is actually going on here. (Lists of pointers make me very suspicious.)

qinheping replied:
I added some file-level comment and an example to explain the codes. The unit test is also an example of how to use the enumerators.
We can't avoid lists of pointers because we need to store the sub-enumerators, which is a list of abstract class, for non-leaf enumerators. Correct me if I am wrong.

tautschnig replied:
I think I'd expect you to wrap these in a class to make sure there is a clearer notation of how to correctly allocate and de-allocate this.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+typedef std::list<enumerator_baset *> enumeratorst;
+
+/// Enumerator that enumerates leaf expressions from a given list.
+/// Leaf expressions are complete expressions with no hole or placeholder
```
What is a "hole," and how does it differ from a "placeholder?"

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+class leaf_enumeratort : public enumerator_baset
+{
+public:
+  explicit leaf_enumeratort(const exprst &leaf_exprs, bool is_none = false)
+    : enumerator_baset(is_none), leaf_exprs(leaf_exprs)
```
I'm not a big fan of default parameters. Do you really need to do this?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+class enumerator_baset
+{
+public:
+  explicit enumerator_baset(bool is_none_enumerator = false)
```
I'm not a big fan of default parameters. Do you really need to do this?

qinheping replied:
Not really. I removed all default parameters.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+  explicit leaf_enumeratort(const exprst &leaf_exprs, bool is_none = false)
+    : enumerator_baset(is_none), leaf_exprs(leaf_exprs)
+  {
+    expected_size = 0;
```
Why not move this to the initializers above? Or perhaps, with C++11, just do it as part of the declaration below.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+  }
+
+protected:
+  /// Size of expected enumerated expressions. The size of a leaf expression
+  /// is 1. So we enumerate only when the `expected_size` is 1.
+  unsigned int expected_size;
```
So this value is really only meaningful for is-one or is-not-one? Perhaps just use a Boolean instead? Should `set_size` check the value being passed?

qinheping replied:
```expected_size``` are also used in `non_leaf_enumeratort`, where it can be not one. So I prefer to keep it integer to make them consistent.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+  /// Size of expected enumerated expressions. The size of a leaf expression
+  /// is 1. So we enumerate only when the `expected_size` is 1.
+  unsigned int expected_size;
+  exprst leaf_exprs;
```
It seems this list is actually constant, only ever populated by the constructor? Is it even right for it to be a list of its own and not a reference? (Maybe yes, for the source object might go out of scope.)

qinheping replied:
I make it a constant list of constant expressions.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+    is_good_partition = [](const partitiont &) { return true; };
+  }
+
+  /// Given a list `enumerators` of enumerators, return the caresian product
```
s/caresian/Cartesian/

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+  /// Order of partitions is considered relevant.
+  std::list<partitiont> get_partitions(unsigned int n, unsigned int k);
+
+  exprst enumerate() override
+  {
+    exprst result = exprst();
+
+    // Enumerate nothing when `expected_size` is too small to be partitioned.
+    if(expected_size - 1 < arity)
+      return result;
+
+    // For every possible partition, set `expected_size` of
+    // each sub-enumerator to be the corresponding component in the partition.
+    for(const auto &partition : get_partitions(expected_size - 1, arity))
+    {
+      if(!is_good_partition(partition))
+        continue;
+
+      set_sub_enumerator_sizes(partition);
+
+      // Compute the cartesian product as result.
+      for(const auto &product_tuple :
+          cartesian_product_of_enumerators(sub_enumerators))
+      {
+        // Optimization: rule out equivalent expressions
+        // using certain equivalence class.
+        // Keep only representation tuple of each equivalence class.
+        if(is_equivalence_class_representation(product_tuple))
+          result.push_back(instantiate(product_tuple));
+      }
+    }
+
+    return result;
+  }
```
Most functions (not just this one) should be moved to the implementation file. This just pollutes the header file, making the API very hard to read.

qinheping replied:
I moved all functions with body more than one line to the implementation file.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+    std::function<bool(const partitiont &)> partition_check)
+    : op_id(op)
+  {
+    is_none_enumerator = false;
```
Redundant? But, really, should be explicitly invoking the parent constructor and get rid of those default parameters.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+    : op_id(op)
+  {
+    is_none_enumerator = false;
+    op_id = op;
```
Already initialised above.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+    is_none_enumerator = false;
+    op_id = op;
+    arity = 2;
+    sub_enumerators = enumeratorst();
```
Unnecessary.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+    op_id = op;
+    arity = 2;
+    sub_enumerators = enumeratorst();
+    sub_enumerators.push_back(enumerator_1.clone());
+    sub_enumerators.push_back(enumerator_2.clone());
```
How about `sub_enumerators({enumerator_1.clone(), enumerator_2.clone()})` in the list of initialisers before the body of this constructor.

qinheping replied:
Good idea.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+    expected_size = new_size;
+  }
+
+  void set_non_exchangable()
+  {
+    is_exchangable = false;
+  }
```
What is this method about?

qinheping replied:
Added a comment to explain exchangeable.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,511 @@
+      is_good_partition);
+  }
+
+  bool is_commutative(const irep_idt op)
+  {
+    return op_id == ID_equal || op_id == ID_plus;
+  }
```
What about multiplication, not-equal, and, or, bitand, bitor, bitxor? You might want to look at, or perhaps even call the functions from src/util/simplify_utils.cpp.

qinheping replied:
Thanks for the reminder. simplify_expr() is now used to rule out more expressions. However, the optimization introduced in this class is still needed to avoid to enumerate redundant expressions.

tautschnig reviewed (changes requested):
I admit to not having read the entire PR, perhaps only the first half of it. As indicated in some comment, I would really appreciate some comments that explain the high-level idea. Various comments of mine hint at cleanup, which would make the code a lot more concise and thereby easier to read.

martin-cs commented:
Maybe I am getting mixed up but I remember @polgreen having some SyGuS / expression synthesis code built with CPROVER.  Maybe with was @pkesseli ?  Am hoping that @peterschrammel remembers...

polgreen commented:
> Maybe I am getting mixed up but I remember @polgreen having some SyGuS / expression synthesis code built with CPROVER. Maybe with was @pkesseli ? Am hoping that @peterschrammel remembers...

We did, but it wasn't enumerating: it was a symbolic encoding using selector variables in a fixed grammar and asking a decision procedure to find assignments to those selector variables that made the expression satisfy some spec. 
The synthesis encoding is here: https://github.com/kroening/fastsynth/blob/master/src/fastsynth/synth_encoding.cpp

martin-cs commented:
Thanks for the super-swift response @polgreen

NlightNFotis commented on src/config.inc:
```diff
@@ -76,7 +76,7 @@ endif
 
 # Detailed version information
-CBMC_VERSION = 5.63.0
+CBMC_VERSION = 5.64.0
```
Please remove this change.

It conflicts with the current version of `src/config.inc`, and releases is something we do on a specific schedule, unless there's an emergency of sorts that requires a hot fix.

NlightNFotis reviewed (dismissed):
Sorry, didn't spend a lot of time reviewing - please ping me when you want to re-review it.

qinheping commented:
> I admit to not having read the entire PR, perhaps only the first half of it. As indicated in some comment, I would really appreciate some comments that explain the high-level idea. Various comments of mine hint at cleanup, which would make the code a lot more concise and thereby easier to read.

Thanks for the comments. I addressed most of the comments and added more explanation and examples.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,364 @@
+  }
+
+  /// Arity n is the number of sub-enumerators
+  arity = sub_enumerators.size();
```
Isn't this the same as `enumerators.size()`?

qinheping replied:
Yes, they are the same.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,364 @@
+    {
+      expr_listt new_tuple;
+      new_tuple.emplace_back(e);
+      result.insert(new_tuple);
```
`result.insert(expr_listt{e})` (though `result.emplace(e)` might even be possible)

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,364 @@
+    {
+      for(const auto &elem : (*it)->enumerate())
+      {
+        expr_listt new_tuple = expr_listt(sub_tuple);
```
`expr_listt new_tuple{sub_tuple};`

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,393 @@
+#ifndef CPROVER_GOTO_INSTRUMENT_SYNTHESIZER_EXPR_ENUMERATOR_H
+#define CPROVER_GOTO_INSTRUMENT_SYNTHESIZER_EXPR_ENUMERATOR_H
+
+#include <util/simplify_expr.h>
```
Why is this needed in a header file?

qinheping replied:
Moved it to the implementation file.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,393 @@
+#include <util/simplify_expr.h>
+#include <util/std_expr.h>
+
+#include <langapi/language_util.h>
```
What is this needed for?

qinheping replied:
It should also be moved to the implementation file, it is for `from_expr`.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,393 @@
+  and the alternatives enumerator E_B is for the production
+  -> 1.
+
+  The idea of the enumeration is to enumerate all possible partitions of sizes,
```
Which "sizes" do you refer to?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,393 @@
+  }
+
+  /// Set the size of expressions enumerated by this enumerator.
+  virtual void set_size(std::size_t new_size) = 0;
```
What is the "size" of an expression? Is it the number of operands?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,393 @@
+  virtual void set_size(std::size_t new_size) = 0;
+
+  virtual expr_sett enumerate() = 0;
+  virtual enumerator_baset *clone() const = 0;
```
When would one need this? You have not deleted the copy constructor, so how is this one different?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,393 @@
+
+  virtual ~enumerator_baset() = default;
+
+  namespacet ns;
```
Should this be `const` and be `protected`?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,393 @@
+protected:
+  /// Size of expected enumerated expressions. The size of a leaf expression
+  /// is 1. So we enumerate only when the `expected_size` is 1.
+  std::size_t expected_size;
```
Why is it even necessary for a `leaf_enumeratort` to track the size?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,393 @@
+
+/// Non-leaf enumerator enumerates expressions of form
+/// f(a_1, a_2, ..., a_n)
+/// where a_i's are leaf expressions enumerated by sub-enumerators.
```
Do they strictly need to be leaf expressions?

qinheping replied:
They are not need to be leaf expressions. It should be "where a_i's are sub-expressions enumerated by sub-enumerators."

nwetzler commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,393 @@
+  and t is some tree composed from nonterminal symbols and constants.
+
+  In the level of productions, we introduce three classes of enumerators:
+  1. `leaf_enumeratort` enumerates `epxrt`-typed expressions. It is used
```
```suggestion
  1. `leaf_enumeratort` enumerates `exprt`-typed expressions. It is used
```

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.h:
```diff
@@ -0,0 +1,393 @@
+  /// \param ns namesapce used for `simplify_expr`.
+  non_leaf_enumeratort(const enumeratorst &enumerators, const namespacet &ns);
+
+  /// Given a list `enumerators` of enumerators, return the Cartesian product
+  /// of expressions enumerated by each enumerator in the list.
+  std::set<expr_listt> cartesian_product_of_enumerators(
```
Can you also explain _why_ you're doing the Cartesian product?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,364 @@
+}
+
+std::list<partitiont>
+non_leaf_enumeratort::get_partitions(std::size_t n, std::size_t k)
```
So as to avoid having to convince myself that your algorithm is correct: how about using https://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation instead?

nwetzler replied:
Seconded.  This took quite a bit of time to understand.

qinheping replied:
Yes, I added an implementation that use the bithacks to enumerate bit permutation. However, such method only works for n <= 64 due to the length of size_t. I rewrote the old implementation using a similar idea of next-bit-permutation to handle n > 64

tautschnig replied:
You could use `mp_integer`, but is it actually realistic to ever want n > 64? Depending on the value of `k` this might get very expensive...

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,364 @@
+  : non_leaf_enumeratort({enumerator_1.clone(), enumerator_2.clone()}, ns),
+    op_id(op)
+{
+  arity = 2;
+  is_good_partition = partition_check;
+  is_exchangeable = exchangeable;
```
Why are these not part of the initializer list, unlike `op_id`?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,364 @@
+
+bool binary_functional_enumeratort::is_commutative(const irep_idt op)
+{
+  return op_id == ID_equal || op_id == ID_plus || op_id == ID_notequal ||
+         op_id == ID_or || op_id == ID_and || op_id == ID_xor ||
+         op_id == ID_bitand || op_id == ID_bitor || op_id == ID_bitxor;
```
Also `ID_mult`

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,364 @@
+    "number of arugment should be 2: " + integer2string(exprs.size()));
+  if(op_id == ID_equal)
+    return equal_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_le)
+    return less_than_or_equal_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_lt)
+    return less_than_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_gt)
+    return greater_than_exprt(exprs.front(), exprs.back());
+  if(op_id == ID_ge)
+    return greater_than_or_equal_exprt(exprs.front(), exprs.back());
```
Do you actually need all of these? For example, `x < y` is the same as `y > x`.

tautschnig reviewed (commented):
A few more comments. This time I have managed to read all the code, and what I find the hardest to comprehend is the business of setting the size (is that the arity?). Why is the current approach the right one?

qinheping commented:
I pushed an update including the following changes:
1. A walkthrough example to illustrate how the enumeration works and why do we need to compute the Cartesian products.
2. Removing all `clone` functions and unnecessary pointers.
3. Removing `set_size` functions and `expected_size`. Now the function `enumerate(std::size_t size)` explicitly take the size of expressions we want to enumerate as argument.
4. Implementing a new `get_partition(n, k)` function using the bithack you suggested. However, it works for only `n` no larger than 64. So I rewrote the old implementation using the same idea of computing next bit permutation to handle case of `n > 64`.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,375 @@
+
+#include <util/simplify_expr.h>
+
+#include <langapi/language_util.h>
```
Is this perhaps a debugging leftover?

qinheping replied:
No, `exprt`s are sorted by comparing their string representation produced by `from_expr`.

tautschnig replied:
Ah! I suggest you use `format` instead (from `util/format_expr.h`).

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,375 @@
+}
+
+std::list<partitiont>
+get_partitions_64(const std::size_t n, const std::size_t k)
```
Nit pick: not all supported platforms have 64-bit size_t types, so the name might be misleading?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,375 @@
+  return result;
+}
+
+/// Construct parition of `n` elements from a bit vector `v`.
```
s/parition/partition/

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,375 @@
+}
+
+/// Construct parition of `n` elements from a bit vector `v`.
+// For a bit vector with ones at positions (computed by `get_ones_pos`)
+// (ones[0], ones[1], ..., ones[k-2]),
+// the corresponding partition is
+// (ones[0], ones[1]-ones[0], ..., ones[k-2]-ones[k-3], n-ones[k-2]).
+partitiont from_bits_to_partition(std::size_t v, std::size_t n)
```
Make all or none of the comment above a Doxygen comment, I'd say.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,375 @@
+    return {};
+
+  // This new implementation works only for n no larger than 64.
+  if(n > 64)
+    return get_partitions_64(n, k);
```
See above: don't hard-code 64 here, use `sizeof` instead.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,375 @@
+  // representation if its sub-expressions are sorted.
+  if(
+    is_exchangeable && is_commutative(op_id) &&
+    from_expr(exprs.front()) > from_expr(exprs.back()))
+    return false;
```
Nit pick: braces around body when the condition spans multiple lines.

qinheping replied:
It is required by check-clang-format.

tautschnig replied:
Apologies for being unclear: yes, up to line 300 this is fine, but then do
```
{
  return false;
}
```

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,375 @@
+{
+  INVARIANT(
+    exprs.size() == 2,
+    "number of arugment should be 2: " + integer2string(exprs.size()));
```
s/arugment/arguments/

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,375 @@
+expr_sett
+recursive_enumerator_placeholdert::enumerate(const std::size_t size) const
+{
+  INVARIANT(factory.productions_map.count(identifier), "No nonterminal found.");
+  alternatives_enumeratort actual_enumerator(
+    factory.productions_map.at(identifier), ns);
```
This entails two lookups of the same element. Perhaps use `.find` and then and `INVARIANT` over the resulting iterator?

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,375 @@
+void enumerator_factoryt::add_placeholder(
+  const recursive_enumerator_placeholdert &placeholder)
+{
+  INVARIANT(
+    nonterminal_set.count(placeholder.identifier) == 0,
+    "Duplicated non-terminals");
+  // The new placeholder (nonterminal) belongs to this factory (grammar).
+  nonterminal_set.emplace(placeholder.identifier);
```
Two lookups. Use `insert` and in the INVARIANT test that `.second` of the result is true.

tautschnig commented on src/goto-instrument/synthesizer/expr_enumerator.cpp:
```diff
@@ -0,0 +1,375 @@
+  const std::string &id,
+  const enumeratorst &enumerators)
+{
+  INVARIANT(
+    nonterminal_set.count(id) == 1,
+    "Cannnot attach enumerators to a non-existing nonterminal.");
+  productions_map.emplace(id, enumerators);
```
As above: avoid two lookups.

qinheping committed:
Introduce grammar-based expression enumerators
